packages/block-editor/src/store/reducer.js

```diff
@@ -497,15 +517,51 @@ const withInnerBlocksRemoveCascade = ( reducer ) => ( state, action ) => {
  * Higher-order reducer which targets the combined blocks reducer and handles
  * the `RESET_BLOCKS` action. When dispatched, this action will replace all
  * blocks that exist in the post, leaving blocks that exist only in state (e.g.
- * reusable blocks) alone.
+ * reusable blocks and blocks controlled by inner blocks controllers) alone.
  *
  * @param {Function} reducer Original reducer function.
  *
  * @return {Function} Enhanced reducer function.
  */
 const withBlockReset = ( reducer ) => ( state, action ) => {
 	if ( state && action.type === 'RESET_BLOCKS' ) {
-		const visibleClientIds = getNestedBlockClientIds( state.order );
+		/**
+		 * A list of client IDs associated with the top level entity (like a
+		 * post or template). It excludes the client IDs of blocks associated
+		 * with other entities, like inner block controllers or reusable blocks.
+		 */
+		const visibleClientIds = getNestedBlockClientIds(
+			state.order,
+			'',
+			state.controlledInnerBlocks
+		);
+
+		// pickBy returns only the truthy values from controlledInnerBlocks
+		const controlledInnerBlocks = Object.keys(
+			pickBy( state.controlledInnerBlocks )
+		);
```

Comment:
I'm not 100% certain about the changes here, I'm not sure we should keep the "controlledInnerBlocks" when we reset parent blocks. Here's an example:

 - Say we have a templatePart { id: 1 }
 - and we reset the templatePart { id: 2 } 

we should reset the blocks, I guess they will be reset anyway by the next render of the block regardless.

So what problems excluding the controlled inner blocks here is solving? Asking cause a long time ago parsed reusable blocks where also kept like this and it created a l to of issues elsewhere.

Also I wonder if there are other selectors like getblockcount that needs to be adapted to exclude controlled blocks.

Reply:
`resetBlocks` will be called in the block editor provider for incoming changes, but they won't include the controlled inner blocks, so we would lose them.

>Also I wonder if there are other selectors like getblockcount that needs to be adapted to exclude controlled blocks.

That's a good question. `getBlockCount` reads from the block order. Do we want to exclude controlled inner blocks here? I think that this and most other block selectors are used for UI that needs the real value. Only saving and change detection stand out.

Reply:
@youknowriad This actually mimics how reset_blocks works with reusable blocks; otherwise I'm not sure why we need this reducer state.

> So what problems excluding the controlled inner blocks here is solving?

If _all_ blocks are removed from the editor during the top-level entity block_reset action (e.g. when you click "undo"), then a template part's controlled inner blocks are removed from the block-editor state and they no longer display. (And I believe if you click save, all the post content becomes blank for the template part too.) That's the specific bug that this resolves: when you click undo, template part blocks disappear from the editor. :D Perhaps there are other ways to accomplish that, but I don't think it makes sense to allow the block_reset action to also control inner blocks that should be controlled by other entities.

> getBlockCount reads from the block order. Do we want to exclude controlled inner blocks here?

I don't think we want to at the moment? If design decides later that we should split word count by entity, maybe? I don't think we need to change it unless we have some motivation :) (Unless this is broken by default)

Like one of the reasons I think this PR works so well is that the block state is not really changed at all. All of the blocks show up exactly where you expect them to. There are just some reducers and selectors that are tweaked to handle controlled inner blocks differently. But if a selector has not received that tweak, then behavior should not change.

Reply:
Yeah, that was my line of thinking as well.

Reply:
I wanted to note that this applies to any parent entity as well. For example, if a template part block calls replace inner blocks, it will remove a child controller’s blocks from the editor. The child will just see that replace inner blocks action as a normal state change (because it is possible for a user to delete all blocks at once), and will therefore delete those blocks from its entity before the block is added back to the state. (At least, this is my working theory, since we have only guarded against a root block reset action so far)